Support slow Start mode in Envoy

nezdolik opened the pull request:
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

Support progressive traffic increase in Envoy, implementation is according to design doc: https://docs.google.com/document/d/1NiG1X0gbfFChjl1aL-EE1hdfYxKErjJ2688wJZaj5a0/edit

Additional Description: Please refer to RFC
Risk Level: Medium
Testing: Done
Docs Changes: Done
Release Notes: Done
Fixes #11050

nezdolik commented:
Planning to use callback mechanism for edf loadbalncer to be aware of which hosts are in slow start mode.

mattklein123 commented:
@nezdolik lmk when you want a first pass on this!

/wait

htuch commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -508,6 +508,18 @@ message Cluster {
+    // Configuration for slow start mode.
+    // [#next-free-field: 3]
+    message SlowStartConfig {
+      google.protobuf.UInt32Value slow_start_window = 1;
```
Please add comment to fields

nezdolik replied:
thanks for review @htuch, i will fix api+docs once PR is in more mature state.

htuch commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -508,6 +508,18 @@ message Cluster {
     }
 
+    enum EndpointWarmingPolicy {
+      NO_WAIT = 0;
```
Please add comment to enum values.

htuch commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -508,6 +508,18 @@ message Cluster {
+      WAIT_FOR_FIRST_PASSING_HC = 1;
+    }
+
+    // Configuration for slow start mode.
```
Can you write some Envoy docs for this and link from here? I'd suggest translating the design doc into RST and then cleaning that up a bit for end users.

sschepens commented:
I'm interested in this, would `WAIT_FOR_FIRST_PASSING_HC`  apply when endpoint health is provided via EDS and no active HC is configured?

nezdolik commented:
@sschepens currently this is the logic for `WAIT_FOR_FIRST_PASSING_HC`:

```
if (host->health() == Upstream::Host::Health::Healthy &&
            !host->healthFlagGet(Host::HealthFlag::PENDING_ACTIVE_HC))
```
This may be not he final version, but currently it checks host health flags. If passive HC keeps those flags up to date it should work.

nezdolik commented:
@mattklein123 @snowp need your initial thoughts on suggested approach for tracking hosts in slow start. The code is still wip and plenty of things will be reworked (eg todos, duplicated code, fix format etc).
Upon creation of edf lb base, we install 2 callbacks that recalculate hosts in slow start on each priority update/cluster membership update. To track hosts in slow start, added collection (set of host ptrs) and time of latest added hosts in slow start. Keeping the time of latest added host in slow start allows to do fast check if any hosts are in slow start.

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -695,17 +695,33 @@ const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts
   priority_set.addPriorityUpdateCb(
-      [this](uint32_t priority, const HostVector&, const HostVector&) { refresh(priority); });
+      [this](uint32_t priority, const HostVector& hosts_added, const HostVector& hosts_removed) {
+        recalculateHostsInSlowStart(hosts_added, hosts_removed);
```
ptal @snowp @mattklein123

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -775,6 +845,24 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
   }
 }
 
+bool EdfLoadBalancerBase::noHostsAreInSlowStart() {
```
ptal @mattklein123 @snowp

mattklein123 replied:
I think this iml makes sense to me at a high level, and avoids complex timers around when hosts end slow start.

cc @antoniovicente @tonya11en who might also be interested in taking a look and brainstorming other options.

mattklein123 commented on include/envoy/upstream/host_description.h:
```diff
@@ -148,6 +148,11 @@ class HostDescription {
+  /**
+   * @return timestamp in milliseconds of when host was created.
+   */
+  virtual std::chrono::milliseconds creationTime() const PURE;
```
I think you want `MonotonicTime` here.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -1241,14 +1241,16 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
         cluster->lbSubsetInfo(), cluster->lbRingHashConfig(), cluster->lbMaglevConfig(),
-        cluster->lbLeastRequestConfig(), cluster->lbConfig());
+        cluster->lbLeastRequestConfig(), cluster->lbConfig(),
+        parent_.thread_local_dispatcher_.timeSource());
```
I might recommend breaking out the creation time portion of this change into its own PR just to keep this meat of this a bit smaller.

mattklein123 commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -388,6 +396,13 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
+      endpoint_warming_policy;
+  const uint32_t slow_start_window;
+  TimeSource& time_source_;
+  absl::node_hash_set<HostSharedPtr> hosts_in_slow_start_;
```
You should be able to use a flat_hash_set here.

nezdolik replied:
it's using `absl::btree_set` now

mattklein123 commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -775,6 +845,24 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
+  // If all hosts are out of the window, we no longer need to track them and therefore we erase
+  // tracked hosts set.
+  if (current_time - latest_host_added_time > slow_start_window_ms) {
+    hosts_in_slow_start_.erase(hosts_in_slow_start_.begin(), hosts_in_slow_start_.end());
```
Just use clear()

mattklein123 commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -775,6 +845,24 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
+    hosts_in_slow_start_.erase(hosts_in_slow_start_.begin(), hosts_in_slow_start_.end());
+    return true;
+  } else {
+    // Second, check if there is any hosts in tracked hosts set.
+    return hosts_in_slow_start_.size() == 0;
```
Can you just see if this is empty at the beginning of the function and return true if empty?

nezdolik replied:
taken care of

mattklein123 reviewed (changes requested):
Thanks for working on this. The shape of this LGTM but I would definitely be interested in hearing from @snowp @antoniovicente @tonya11en if they have other impl ideas. Thank you!

/wait

nezdolik commented:
Just realised that storing only time of latest added host will not work, for example in case host is added to the cluster and then immediately removed. It needs to be more complex data structure, that supports ordering by time, querying for latest time and lookups by host.

mattklein123 commented:
> Just realised that storing only time of latest added host will not work, for example in case host is added to the cluster and then immediately removed. It needs to be more complex data structure, that supports ordering by time, querying for latest time and lookups by host.

Assuming we stick with the high level approach, I think you could probably use `absl::btree_set` with a heterogeneous lookup implementation that allows storing a host and sorting on last addition time.

nezdolik commented:
> 
> Assuming we stick with the high level approach, I think you could probably use `absl::btree_set` with a heterogeneous lookup implementation that allows storing a host and sorting on last addition time.

Thanks for the tip, this is what i was looking for. For now will focus on submitting separate PR with adding "creation time" to all Host entities.

nezdolik commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -338,7 +362,9 @@ TEST_P(LoadBalancerBaseTest, GentleFailover) {
   // P=0 should receive 4/(4+6)=40% of traffic
   // P=1 should receive 6/(4+6)=60% of traffic
-  updateHostSet(host_set_, 4 /* num_hosts */, 0 /* num_healthy_hosts */, 0 /* num_degraded_hosts */,
+  updateHostSet(host_set_, 4 /* num_hosts */, 0 /* num_healthy_hosts */, 0 /* num_degraded_hosts
```
oh some weird formatting

nezdolik commented:
@mattklein123 @snowp please take a look. If this approach is good enough in RR i will add slow start to LR load balancer as well. 
Left to do:
* slow start & tests in LR
* docs
* add tests for unsupported slow start config for other LB types

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -705,17 +705,41 @@ const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts
       [this](uint32_t priority, const HostVector&, const HostVector&) { refresh(priority); });
+  priority_set.addMemberUpdateCb([this](const HostVector& hosts_added, const HostVector&) -> void {
+    if (slow_start_enabled_) {
+      recalculateHostsInSlowStart(hosts_added);
```
hosts in slow start are now recalculated in membership update callback + in refresh upon priority update (to catch health state change)

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -410,6 +420,29 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
+  struct orderByCreateDateDesc {
+    bool operator()(const HostSharedPtr l, const HostSharedPtr r) const {
+      return (l->creationTime() < r->creationTime()) ||
+             (l->address()->asString() < r->address()->asString());
```
experimented with comparator a little bit and this one is optimal for ordering+unique host retrieval/removal (relies on host addresses (hostAndPort) being unique)

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -724,20 +748,42 @@ void EdfLoadBalancerBase::initialize() {
 }
 
+void EdfLoadBalancerBase::recalculateHostsInSlowStart(const HostVector& hosts) {
+  auto current_time = time_source_.monotonicTime();
```
now as hosts_in_slow_start (with new comparator) support removal by single host, could process hosts_removed and remove those from slow start collection.

mattklein123 commented:
@nezdolik can you work on the docs (especially the API docs) and also get CI passing? Then I can take a look. Thank you!

/wait

nezdolik commented:
added API docs, user docs and fixed build

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -504,6 +511,22 @@ message Cluster {
+      // Size of slow start window in seconds, default value is 0.
+      // If set to value greater than 0, the newly created host remains in slow start mode starting from it's creation time
+      // for the duration of slow start window.
+      google.protobuf.UInt32Value slow_start_window = 1 [(validate.rules).uint32 = {gte: 0}];
```
If the default is 0, this can just be a basic type?

nezdolik replied:
My comment was actually wrong, default value in code is 1.0.

nezdolik replied:
changed

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -436,11 +436,18 @@ message Cluster {
 
+    // Refer to :ref:`endpoint warming policy types <arch_overview_load_balancing_slow_start_endpoint_warming_policy_types>`
+    // for an explanation on each policy type.
+    enum EndpointWarmingPolicy {
+      NO_WAIT = 0;
+      WAIT_FOR_FIRST_PASSING_HC = 1;
+    }
```
inline this into `SlowStartConfig`?

nezdolik replied:
think the initial suggestion was to reuse this message type across other mechanisms in Envoy. We could make it top level message later if it will be actually used anywhere else.

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -504,6 +511,22 @@ message Cluster {
+      EndpointWarmingPolicy endpoint_warming_policy = 2
+          [(validate.rules).enum = {defined_only: true}];
+
+      core.v3.RuntimeDouble time_bias = 3;
```
docs?

nezdolik replied:
oh, somehow missed that one

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -545,6 +568,9 @@ message Cluster {
     // Common Configuration for all consistent hashing load balancers (MaglevLb, RingHashLb, etc.)
     ConsistentHashingLbConfig consistent_hashing_lb_config = 7;
+
+    // Configuration for slow start mode.
```
Say what happens if not set? No slow start?

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -436,11 +436,18 @@ message Cluster {
+    // Refer to :ref:`endpoint warming policy types <arch_overview_load_balancing_slow_start_endpoint_warming_policy_types>`
+    // for an explanation on each policy type.
+    enum EndpointWarmingPolicy {
+      NO_WAIT = 0;
+      WAIT_FOR_FIRST_PASSING_HC = 1;
```
I would still write short descriptions here even if the arch docs cover this.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
+Slow start mode
+===============
+
+Slow start mode is a configuration setting in Envoy to progressively increase amount of traffic for newly spawned service instances.
```
```suggestion
Slow start mode is a configuration setting in Envoy to progressively increase amount of traffic for newly added upstream endpoints.
```

mattklein123 replied:
(Similar changes to s/instances/endpoints, etc. elsewhere)

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
+With no slow start enabled Envoy would send a proportional amount of traffic to new instances.
+This could be undesirable for services that require warm up time to serve full production load and could result in request timeouts, loss of data and deteriorated user experience.
+
+Slow start mode is a weight-adjustment mechanism in number of load balancer types and can be configured per cluster basis. 
```
Can't quite parse. Can you rephrase?

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
+
+To reiterate, host enters slow start mode when:
+  * For :ref:`NO_WAIT<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.EndpointWarmingPolicy.NO_WAIT>` endpoint warming policy, immediately if it's cluster membership duration is within slow start window.
+  * For :ref:`WAIT_FOR_FIRST_PASSING_HC<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.EndpointWarmingPolicy.WAIT_FOR_FIRST_PASSING_HC>` endpoint warming policy, upon passing an active healthcheck, given that it's cluster membership duration is within slow start window.
```
I can't quite tell from reading this if you mean that the duration is implicitly extended by the duration until the first passing HC? Can you rephrase/expand?

nezdolik replied:
How is currently implemented:
With `WAIT_FOR_FIRST_PASSING_HC` policy, endpoint enters slow start if its creation time is within slow start window and if it has passed an active HC. Duration is not implicitly extended. Host exits slow start if it gets marked unhealthy (via priority update CB). Host could further re-enter slow start mode if it passes HC again and if it's creation date is still within slow start window. Name `WAIT_FOR_FIRST_PASSING_HC` is misleading, should be `WAIT_FOR_PASSING_HC` instead. This behaviour has not been discussed in initial RFC, i've came up with it during implementation. Happy to discuss more.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
+Host exists slow start mode when:
+  * It leaves the cluster.
+  * It's cluster membership duration is greater than slow start window.
+  * For :ref:`WAIT_FOR_FIRST_PASSING_HC<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.EndpointWarmingPolicy.WAIT_FOR_FIRST_PASSING_HC>` endpoint warming policy, if host's health status changes to anything rather than `Healthy`.
```
I'm wondering if this makes sense? Should we still adhere to the duration if a host becomes unhealthy? It seems like maybe this could be simplified somewhat? Maybe I don't fully understand the use case.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
+  * For :ref:`WAIT_FOR_FIRST_PASSING_HC<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.EndpointWarmingPolicy.WAIT_FOR_FIRST_PASSING_HC>` endpoint warming policy, if host's health status changes to anything rather than `Healthy`.
+
+Below is example of how requests would be distributed across hosts with Round Robin Loadbalancer, slow start window of 10 seconds, `NO_WAIT` ednpoint warming policy and 0.5 time bias.
+Host H1 has statically configured initial weight of X and host H2 weight of Y, the actual numerical values are of no significance for this example.
```
s/Host/Endpoint and switch to E1, etc.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
++-------------+----------------+------------+------------+-----------+----------+-------------+
+
+
+.. _arch_overview_load_balancing_slow_start_endpoint_warming_policy_types:
+
+Endpoint warming policy types
+-----------------------------
+
+Endpoint warming policy defines conditions for host to enter slow start mode.
+
+No Wait
+^^^^^^^
+
+If configured, host would enter slow start immediately.
+
+Wait For First Passing Healthcheck
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If configured, host would enter slow start upon having passed an active healthcheck.
```
I think this can be moved back to the API docs, assuming we still need the policy after discussion.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,66 @@
+  * It's cluster membership duration is greater than slow start window.
+  * For :ref:`WAIT_FOR_FIRST_PASSING_HC<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.EndpointWarmingPolicy.WAIT_FOR_FIRST_PASSING_HC>` endpoint warming policy, if host's health status changes to anything rather than `Healthy`.
+
+Below is example of how requests would be distributed across hosts with Round Robin Loadbalancer, slow start window of 10 seconds, `NO_WAIT` ednpoint warming policy and 0.5 time bias.
```
This is the first time you have discussed "time bias." Can you introduce this concept?

mattklein123 reviewed (changes requested):
Very cool to see this coming together. Flushing out some API/doc comments from the first pass. Thank you!

/wait

nezdolik commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,71 @@
+To reiterate, endpoint enters slow start mode when:
+  * For :ref:`NO_WAIT<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.NO_WAIT>` endpoint warming policy, immediately if it's cluster membership duration is within slow start window.
+  * For :ref:`WAIT_FOR_FIRST_PASSING_HC<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.WAIT_FOR_FIRST_PASSING_HC>` endpoint warming policy, if it's cluster membership duration is within slow start window and endpoint has passed an active healthcheck. 
+    If endpoint does not pass an active healcheck during entire slow start window (since it has been added to upstream cluster), then it never enter slow start mode.
```
s/never enter/never enters

mattklein123 replied:
I still don't quite understand why we need this configuration. Can we just say that if active HC is configured slow start starts from the time of first passing HC, otherwise it starts immediately? It seems like we don't need this config immediately?

nezdolik commented:
@mattklein123 applied review comments and expanded on impl details for `WAIT_FOR_FIRST_PASSING_HC` case. Will proceed with adding tests for unsupported slow start mode config in various LB types (unsupported in all except RR and LR).

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -507,6 +507,42 @@ message Cluster {
+        // Endpoint would enter slow start immediately.
+        NO_WAIT = 0;
+
+        // Endpoint would enter slow start upon passing first active healthcheck.
```
Say what happens if health checks are not configured?

nezdolik replied:
Should we keep mentioning "active" HC? As passive HC could also flip/ clear host health flag and it would be considered healthy and satisfy this endpoint warming policy.

mattklein123 replied:
Per other comments, IMO we should remove this policy entirely from the API for now. I think the way it should work is:
1) If no active HC configured, enter slow start immediately
2) If active HC configured, enter slow start after the first transition to healthy

nezdolik replied:
Alright, i need to inspect the code on "how to check if active HC is configured per cluster".

nezdolik replied:
Looks like i could use criteria `host->healthChecker() != null` on any host (as hc is configured per cluster so either all of hosts have hc configured or none) in priority update callback.

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -507,6 +507,42 @@ message Cluster {
+      // Size of slow start window in seconds, defaults to 0.
+      // If set to value greater than 0, the newly created host remains in slow start mode starting from it's creation time
+      // for the duration of slow start window.
+      uint32 slow_start_window = 1 [(validate.rules).uint32 = {gte: 0}];
```
validation not needed, though what would be the use case for setting this to 0?

nezdolik replied:
this would effectively disable slow start. Right, uint32 cannot be negative :)

mattklein123 replied:
My point is >= is guaranteed, so I would either make it > or remove it. I think you want > 0? The text above about defaulting to 0 doesn't make sense I don't think. I think the lack of a SlowStartConfig at all is enough to enable/disable slow start, so for this param I would just make it > 0 and update the text.

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -507,6 +507,42 @@ message Cluster {
+      }
+
+      // Size of slow start window in seconds, defaults to 0.
+      // If set to value greater than 0, the newly created host remains in slow start mode starting from it's creation time
```
```suggestion
      // If set to value greater than 0, the newly created host remains in slow start mode starting from its creation time
```

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -507,6 +507,42 @@ message Cluster {
+      // The smaller the time bias is, the less traffic would be sent to endpoint that is within slow start window.
+      // The value of time bias should be greater than or equal to 0.0 and less than or equal to 1.0.
+      // If slow start window and time bias are configured, effective weight of an endpoint would be scaled with time bias:
+      // `weight = load_balancing_weight * time_bias`
```
Can you clarify how time itself factors into this calculation or the overall calculation? It's not completely clear to me.

mattklein123 replied:
Ping

nezdolik replied:
@mattklein123 would this be better?
```
If slow start window and time bias are configured, effective weight of an endpoint would be scaled with time bias during the duration of slow start window:
`weight = load_balancing_weight * time_bias`
Once duration of slow start window elapses, hosts exists slow start and its weight is no longer being scaled with time bias parameter, the regular weight is applied.
```
I may not be understanding what else  is missing in the description. It is as simple as apply time bias to weight during slow start window and not apply the bias outside of the window.

mattklein123 replied:
What I'm asking for is to update the formula to account for time. Isn't time somehow related to the final weight?

nezdolik replied:
The current implementation only checks if host creation time is within time window or not and if it is it then multiplies weight with bias, eg (shortened version):
```
      auto host_weight = hostWeight(*host);
      if (host_create_duration <= slow_start_window_ ) {

        time_bias_ = time_bias_runtime_ != nullptr ? time_bias_runtime_->value() : time_bias_;

        if (time_bias_ < 0.0) {
          ENVOY_LOG(warn, "upstream: invalid time bias supplied (runtime key {}), using 1.0",
                    time_bias_runtime_->runtimeKey());
          time_bias_ = 1.0;
        }
        if (time_bias_ > 0.0) {
          host_weight *= time_bias_;
        }
      }
```

mattklein123 replied:
I see. That's not typically how slow start works. Usually the weight ramps over the slow start window. I would suggest updating the formula to account for time?

nezdolik replied:
Yes, good point. I was not sure myself, as current formula would reduce amount of traffic sent to endpoint as time within slow start window is progressing. Will fix the formula.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,71 @@
+This could be undesirable for services that require warm up time to serve full production load and could result in request timeouts, loss of data and deteriorated user experience.
+
+Slow start mode is a mechanism that affects load balancing weight of upstream endpoints and can be configured per upstream cluster. 
+Currently, slow start is supported in Round Robin and Least Request load balancer types.
```
Do we fail the config if configured on other types currently?

nezdolik replied:
have not wrapped up this part of code yet, but RFC suggests that loading of config should fail if slow start is configured for unsupported lb type.

nezdolik replied:
this is no longer a case, as slow start config is included into specific lb type config (instead of being part of common lb config)

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,71 @@
+Currently, slow start is supported in Round Robin and Least Request load balancer types.
+
+Users can specify a :ref:`slow start window parameter<envoy_v3_api_field_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.slow_start_window>` (in seconds), so that if endpoint “cluster membership duration" (amount of time since it has joined the cluster) is within the configured window, it enters slow start mode, given that the endpoint satisfies endpoint warming policy. 
+During slow start window, load balancing weight of a particular endpoint will be scaled with :ref:`time bias parameter<envoy_v3_api_field_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.time_bias>`, e.g.:
+`weight = load_balancing_weight * time_bias`.
```
Same question as above, how does time itself factor in?

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,71 @@
+Endpoint could also exit slow start mode in case it leaves the cluster.
+
+To reiterate, endpoint enters slow start mode when:
+  * For :ref:`NO_WAIT<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.NO_WAIT>` endpoint warming policy, immediately if it's cluster membership duration is within slow start window.
```
```suggestion
  * For :ref:`NO_WAIT<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.NO_WAIT>` endpoint warming policy, immediately if its cluster membership duration is within slow start window.
```

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,71 @@
+
+To reiterate, endpoint enters slow start mode when:
+  * For :ref:`NO_WAIT<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.NO_WAIT>` endpoint warming policy, immediately if it's cluster membership duration is within slow start window.
+  * For :ref:`WAIT_FOR_FIRST_PASSING_HC<envoy_v3_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.WAIT_FOR_FIRST_PASSING_HC>` endpoint warming policy, if it's cluster membership duration is within slow start window and endpoint has passed an active healthcheck. 
```
Same question about what if HC is not configured.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,71 @@
++-------------+--------------------+------------+------------+-----------+----------+-------------+
+
+
+.. _arch_overview_load_balancing_slow_start_endpoint_warming_policy_types:
+
+Endpoint warming policy types
+-----------------------------
+
+Endpoint warming policy defines conditions for host to enter slow start mode.
+
+No Wait
+^^^^^^^
+
+If configured, endpoint would enter slow start immediately.
+
+Wait For First Passing Healthcheck
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If configured, endpoint would enter slow start upon having passed an active healthcheck.
```
This can get deleted and just be in the API docs

mattklein123 reviewed (changes requested):
Thanks flushing some more API/doc comments.

/wait

nezdolik commented:
Removed the concept of endpoint warming policy and applied most of api/doc review comments. Now need to fix tests.

zyfjeff commented on source/common/upstream/subset_lb.cc:
```diff
@@ -38,9 +39,14 @@ SubsetLoadBalancer::SubsetLoadBalancer(
+      time_source_(time_source) {
   ASSERT(subsets.isEnabled());
 
+  if (common_config.has_slow_start_config()) {
```
Subset + Roundrobin /LeastRequest is supported.

nezdolik replied:
fixed

zyfjeff commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -695,17 +695,38 @@ const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts
   // https://github.com/envoyproxy/envoy/issues/2874).
   priority_update_cb_ = priority_set.addPriorityUpdateCb(
       [this](uint32_t priority, const HostVector&, const HostVector&) { refresh(priority); });
+  member_update_cb_ = priority_set.addMemberUpdateCb(
```
Is it necessary to add all hosts to the `hosts_in_slow_start_` list when it is first started?

nezdolik replied:
@zyfjeff not sure I understand the question. Could you rephrase?

zyfjeff replied:
When EDS pushes all the hosts for the first time, they will be in slow mode. Is this necessary?

nezdolik replied:
@zyfjeff hosts should enter slow start mode as soon as they join the cluster (if no active HC is configured per cluster). If active HC is configured they enter slow start upon passing first HC.

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -773,6 +831,44 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
+
+bool EdfLoadBalancerBase::adheresToEndpointWarmingPolicy(const Host& host) {
+  // Enter slow start innediately.
+  if (host.healthChecker().isNull()) {
```
am now thinking that simplier way could have been just to check `host.health() == Upstream::Host::Health::Healthy`. It works for both cases, if no HC is configured then host is always healthy otherwise active HC would update host health flag accordingly.

nezdolik commented:
This is ready for next round of review. The only part that is missing is tests for unsupported slow start config in other Lb types.

mattklein123 reviewed (changes requested):
Few more doc/API comments. Please also check CI for the build. Thank you!

/wait

nezdolik commented:
@mattklein123 @antoniovicente @tonya11en please review my suggestion for slow start formula: https://docs.google.com/document/d/17BHC5k5XiJVMIk-B0ZY_tnxjHpH7B-8eppeSJaFxtsk/edit?usp=sharing

nezdolik commented:
Formula and docs updated according to suggestion. Need to update the tests.

mattklein123 commented:
I'm going to be out on leave for a while, so i'm assigning over to @tonya11en to help finish reviewing this and land it. Excited to see this land!

tonya11en commented:
I'm moving the formula discussion over here from the google doc. You had a formula that looks like:

`f(x) = C * time_bias * ( (1 / slow_start_window_value) * creation_duration)`

I think this would work for scenarios where you need linear ramp-up, but I'm not convinced that's the majority of cases. Thinking about a basic cache use-case, you're attempting to reduce the impact of the cache misses over time by sending less traffic to the cold caches, right? Assuming a fixed working set size, the probability of a cache miss will be non-linear (regardless of whether the access pattern is uniform/zipf/pareto).

I think this is fundamentally trying to just ramp up from 0% -> 100% traffic over some time interval, so it looks just like the admission control rejection probability curves, but backwards:
![image](https://user-images.githubusercontent.com/1109246/115069303-3cff7100-9ea8-11eb-98b7-d268e2ecb84d.png)

I think you can adapt that formula and give folks an option of how aggressive they want the ramp-up to be. Just change the x-axis to time and the rejection probability to the % of the target weight.

@nezdolik I'm interpreting `time_bias` as just the slope of this linear ramp-up, so I assume you have this linear fn:
```
new_weight = weight * time_bias * (slow_start_time / time_since_start)
```

We can tweak this to allow for the curve to be exponential, polynomial, or linear depending on an "aggression" value to account for the different types of workloads:
```
new_weight = weight * time_bias * (slow_start_time / time_since_start) ^ (1 / aggression)
```

** I need to run to my desktop to build the graphs, so I will update this post shortly. **

So now we can maintain the linear ramp-up you're proposing here by just setting `aggression=1`, but we also have the option of having a more natural exponential curve which mimics the behavior of the cache warming use-cases. WDYT?

tonya11en commented:
/wait-any

nezdolik commented:
Thanks @tonya11en for review and great suggestions! I was aiming for linear ramp up and did not think much about functions with faster growth speed (eg polynomial. exponential). Will try to test various curves and incorporate suggestions this week.

tonya11en commented:
Awesome, I'll stand by for the results.

/wait-any

arikachen commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,57 @@
+Slow start mode is a mechanism that affects load balancing weight of upstream endpoints and can be configured per upstream cluster. 
+Currently, slow start is supported in Round Robin and Least Request load balancer types.
+
+Users can specify a :ref:`slow start window parameter<envoy_v3_api_field_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.slow_start_window>` (in seconds), so that if endpoint “cluster membership duration" (amount of time since it has joined the cluster) is within the configured window, it enters slow start mode. 
```
```“cluster```

nezdolik replied:
good catch, hard to spot

nezdolik commented:
Updated API, docs and code with new formula. Need to update tests as well.

tonya11en reviewed (changes requested):
I realize you're still working things out with this patch, but I took a quick look at the calculation and wanted to get you some comments isolated to that. This is really cool!

/wait

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -442,7 +476,33 @@ class RoundRobinLoadBalancer : public EdfLoadBalancerBase {
+          host.health() == Upstream::Host::Health::Healthy) {
+
+        time_bias_ = time_bias_runtime_ != nullptr ? time_bias_runtime_->value() : 1.0;
+        aggression_ = aggression_runtime_ != nullptr ? aggression_runtime_->value() : 1.0;
```
forgot about it, but now is fixed

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -442,7 +476,33 @@ class RoundRobinLoadBalancer : public EdfLoadBalancerBase {
+        time_bias_ = time_bias_runtime_ != nullptr ? time_bias_runtime_->value() : 1.0;
+        aggression_ = aggression_runtime_ != nullptr ? aggression_runtime_->value() : 1.0;
+
+        if (time_bias_ < 0.0) {
```
i've opted for second option, as it could be hard to justify a very specific number that would work for clusters with various traffic patterns

nezdolik commented:
@tonya11en thanks for review, i've applied review comments and finished all test for Round Robin.
What is left:
- add slow start support in Least Request & write tests.
- update last column of slow start table in docs (to account for aggression and time factor).
- add tests for unsupported LB types.

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -741,6 +791,8 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
       // at which point it is reinserted into the EdfScheduler with its new
       // weight in chooseHost().
       scheduler.edf_->add(hostWeight(*host), host);
+      std::cerr << "***Adding weight " << hostWeight(*host) << "for host "
```
this is to be removed

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -415,6 +424,30 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
+  const std::chrono::milliseconds slow_start_window_;
+  double time_bias_{1.0};
+  double aggression_{1.0};
+  const std::unique_ptr<Runtime::Double> time_bias_runtime_;
+  const std::unique_ptr<Runtime::Double> aggression_runtime_;
```
i've tried to follow the pattern of active_request_bias_runtime_, should we then change all 3 of them?

nezdolik commented:
Added implementation and tests for slow start in Least Request. Remaining:
* update last column of slow start table in docs (to account for aggression and time factor).
* add tests for unsupported LB types

tonya11en commented:
> @tonya11en this is ready for review

I'll take a look today. Thanks!

wbpcode commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -442,7 +480,13 @@ class RoundRobinLoadBalancer : public EdfLoadBalancerBase {
     peekahead_index_ = 0;
   }
-  double hostWeight(const Host& host) override { return host.weight(); }
+  double hostWeight(const Host& host) override {
+    if (slow_start_enabled_) {
+      return applySlowStartFactor(host.weight(), host);
+    }
+    return host.weight();
+  }
```
In the current implementation, the refresh method is not actively called after all nodes have left slow start mode. So even if all the final weights are the same, we will continue to use the EDF scheduler and keep on calculating the time factor which is already meaningless. 

In addition, assuming that the endpoints themselves have different weights, the EDF scheduler will always be used. Then after the slow start mode is over, the time factor calculation will continue even if the refresh method is actively called, which is also an unnecessary overhead. Alternatively, we can add an extra status bit to the endpoint to stop the time factor calculation after the node leaves the slow start mode.

wbpcode reviewed (changes requested):
I'm very happy to see this feature realized. Thanks, It's a great work and make sense in many practices.
But in the concrete implementation, here are some minor comments.

tonya11en reviewed (changes requested):
These docs are great. Thanks for being so thorough there!

I'm overall happy with how the actual slow-start function is turning out, so thanks for entertaining my earlier comments on the aggression term. That being said, I have some fundamental concerns about the time bias value:

1. When you have both a time _bias_ (slope of the ramp-up) and a time _factor_ (essentially length of the x-axis), they're serving the same purpose. Increasing both the time bias and slow-start time by the same amount results in a no-op. They're really just the same term, right? If I'm way off base here, let me know.
2. Values >1.0 for the time bias would result in a temporary increase in the host weight beyond what it was configured for, right? For instance, a host with weight 10 and a time bias of 2.0 (agression=1.0) would cause the adjusted host weight to reach its normal value (10) halfway through the slow-start window. Beyond that it keeps increasing until 20 and then drops to 10 when it exits the slow-start window and you stop applying the slow-start factor. That behavior is going to be problematic.

One of the reasons I pitched that aggression exponent a while back is that it's normalized by the slow-start window, so you can never really exceed the original host weight. It's just fiddling with the curve on the way there. The time bias is actually changing the value of the weight itself.. let me know your thoughts.

The other comments are pretty minor. Please feel comfortable pushing back on any of it. I don't feel too strongly about them.

Thanks again for doing this and going through all the back-and-forth. When this is finished, it's going to be an awesome feature.

tonya11en commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -415,6 +428,30 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
+  // Not meant to check if given host is in slow start as check relies on current time.
+  // That check could be done inline by comparing the diff of current time and host creation date
+  // against slow start window and by checking if hosts adheres to endpoint warming policy.
+  std::shared_ptr<absl::btree_set<HostSharedPtr, OrderByCreateDateDesc>> hosts_in_slow_start_;
```
+1 to a timer and bool state. I think that's pretty elegant.

nezdolik replied:
@wbpcode Just to make sure I understand the suggestion. Keeping only one timer of latest added host is not enough to reliably track if any host is in slow start . Is the suggestion to replace btree with a collection of timers and one bool variable, in that case what is the need for bool variable? Or did you mean storing a map of <timer, bool_state>, in that case the presence of timer in the map can be an indication if some host is in slow start, no need to store the bool state as well. (once timer expires, it gets removed from collection)

wbpcode replied:
@nezdolik 🤔 I think it is enough to just create a timer for the last host added. When this timer expires, all hosts must be outside the slow start time window.
Can you explain why you think one timer is not enough?

nezdolik replied:
@wbpcode in my first implementation of slow start i was using timestamp of latest added host, but then realised that it will not cover some use cases, for example:
![Image from iOS (7)](https://user-images.githubusercontent.com/14430883/126291494-3c67e1b1-374c-41ee-9c67-a90c09e71944.jpg)

nezdolik replied:
https://github.com/envoyproxy/envoy/pull/13176#issuecomment-704148676

wbpcode replied:
@nezdolik I understand. But I think that in this case you said, the price you pay is only in a slow_start_window time, the cluster is mistaken for being in a slow start state, and some additional calculations are required. This is not unacceptable overhead. And the implementation is cleaner and simpler.
To be honest, envoy is complex enough, and it’s not a bad thing to be simple and effective.

nezdolik replied:
i now need to reread impl code again after a 2 month pause. The price you pay would be that in unlucky case Envoy would consider there are no hosts in slow start even if there are some (like host A in example above), edf creation would be skipped and the weights will not be recalculated. wdyt @wbpcode

nezdolik commented:
Thanks for great and thorough review comments, they all make sense to me, am working on updating the PR.

ppmspetrovic commented:
A keep-alive comment is presumably appropriate, given this is live ongoing work.

gleichda commented:
@nezdolik as this is quite a bit older now already any update on the update of the PR? Really looking forward to it. Or anywhere I could support you (even though my C(++) Knowledge is very limited)

nezdolik commented:
> These docs are great. Thanks for being so thorough there!
> 
> I'm overall happy with how the actual slow-start function is turning out, so thanks for entertaining my earlier comments on the aggression term. That being said, I have some fundamental concerns about the time bias value:
> 
> 1. When you have both a time _bias_ (slope of the ramp-up) and a time _factor_ (essentially length of the x-axis), they're serving the same purpose. Increasing both the time bias and slow-start time by the same amount results in a no-op. They're really just the same term, right? If I'm way off base here, let me know.
> 2. Values >1.0 for the time bias would result in a temporary increase in the host weight beyond what it was configured for, right? For instance, a host with weight 10 and a time bias of 2.0 (agression=1.0) would cause the adjusted host weight to reach its normal value (10) halfway through the slow-start window. Beyond that it keeps increasing until 20 and then drops to 10 when it exits the slow-start window and you stop applying the slow-start factor. That behavior is going to be problematic.
> 
> One of the reasons I pitched that aggression exponent a while back is that it's normalized by the slow-start window, so you can never really exceed the original host weight. It's just fiddling with the curve on the way there. The time bias is actually changing the value of the weight itself.. let me know your thoughts.
> 
> The other comments are pretty minor. Please feel comfortable pushing back on any of it. I don't feel too strongly about them.
> 
> Thanks again for doing this and going through all the back-and-forth. When this is finished, it's going to be an awesome feature.

@tonya11en I agree that `time_bias` and `aggression` serve same purpose and it may be confusing for users. We should not complicate configuration unless is really needed. Let's stick with aggression only.

Am currently learning how timers work in Envoy to swap existing implementation with cleaner one. (as per @wbpcode suggestion). Current solution was way easier to implement, that's why i did not went with timers in first place.

nezdolik commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -508,6 +508,34 @@ message Cluster {
+      // Size of slow start window in seconds, should be set to value greater than zero.
+      // If set, the newly created host remains in slow start mode starting from its creation time
+      // for the duration of slow start window.
+      uint32 slow_start_window = 1 [(validate.rules).uint32 = {gt: 0}];
```
there is no smart reason for it

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -714,20 +741,43 @@ void EdfLoadBalancerBase::initialize() {
+    // case EDF creation is skipped. When all original weights are equal and no hosts are in slow
+    // start mode we can rely on unweighted host pick to do optimal round robin and least-loaded
+    // host selection with lower memory and CPU overhead.
+    if (hostWeightsAreEqual(hosts) && noHostsAreInSlowStart()) {
```
good catch

nezdolik replied:
tco

tonya11en commented:
> @tonya11en I agree that `time_bias` and `aggression` serve same purpose and it may be confusing for users. We should not complicate configuration unless is really needed. Let's stick with aggression only.
> 
> 
> 
> Am currently learning how timers work in Envoy to swap existing implementation with cleaner one. (as per @wbpcode suggestion). Current solution was way easier to implement, that's why i did not went with timers in first place.

@nezdolik I can send some patches to so it doesn't totally fall on you (maybe a PR to your branch?). I'll admit that I put in a tall order here. Let me know if that would help, otherwise I'll stand by for the next revision.

nezdolik commented:
@tonya11en thank you for offering help, i already started applying some review comments, but i can tag you in some i have not started yet. The reason for my slowness that am also working on another big [PR](https://github.com/envoyproxy/envoy/pull/15707) and juggling those 2 projects is not going as smooth as i would wish.

nezdolik commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,84 @@
+Below is example of how requests would be distributed across endpoints in same priority with Round Robin Loadbalancer, slow start window of 60 seconds, no active healcheck, 0.5 time bias and 1.0 aggression.
+Endpoints E1 and E2 have statically configured initial weight of X, the actual numerical value is of no significance for this example.
+
++-------------+--------------------+------------+------------+-----------+----------+-------------+
```
have not started this one

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -571,13 +621,19 @@ class LeastRequestLoadBalancer : public EdfLoadBalancerBase,
       : ZoneAwareLoadBalancerBase(priority_set, local_priority_set, stats, runtime, random,
-                                  common_config) {}
+                                  common_config) {
+    if (common_config.has_slow_start_config()) {
```
have not started this one

nezdolik commented:
* factored out slow start config from common lb config and referenced it only from RR and LR types
* switched from using std::unique_ptr to absl::optional for runtime biases (aggression, active request bias)
* got rid of excessive time_bias computation when there are no hosts in slow start anymore
* fixed bunch of tests, but more fix is needed

nezdolik commented:
* Fixed test
* Switched to using diagram instead of table for slow start in action practical example

tonya11en commented:
Sorry for the delay.. I set aside some time to take a look at this tomorrow.

nezdolik commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -485,17 +529,17 @@ class RoundRobinLoadBalancer : public EdfLoadBalancerBase {
  */
-class LeastRequestLoadBalancer : public EdfLoadBalancerBase,
-                                 Logger::Loggable<Logger::Id::upstream> {
+class LeastRequestLoadBalancer : public EdfLoadBalancerBase {
```
@tonya11en since we are modifying `hostWeight` function in both RR and LR, it would be odd/inconsistent to only subclass LR type. I agree that it would make code more readable, but it also brings code complexity in other part of codebase. In code part where LBs are created we will need to check if slow start config is defined and then create subclass types of RR/LR instead of regular types. I would leave this code as it is, but what do you think?

tonya11en replied:
Yeah, that makes sense to me.

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -833,6 +910,36 @@ HostConstSharedPtr EdfLoadBalancerBase::chooseHostOnce(LoadBalancerContext* cont
+    auto time_factor = static_cast<double>(std::max(std::chrono::milliseconds(1).count(),
+                                                    host_create_duration.count())) /
+                       slow_start_window_.count();
+    return host_weight * time_bias_ * applyAggressionFactor(time_factor);
```
so far i've added recommendations for when to not to use slow start mode (based on this convo) into user docs, lmk if you think more drastic approach would make sense (i think is fine to have it in user docs)

nezdolik commented:
Here is preview of documentation and generated graphs:
Part 1
![Screenshot 2021-09-10 at 13 16 56](https://user-images.githubusercontent.com/14430883/132845693-d41a1c10-f0fa-47aa-ade1-93f324d31e37.png)
Part 2
![Screenshot 2021-09-10 at 13 17 15](https://user-images.githubusercontent.com/14430883/132845703-2fdb0285-59ba-4e06-a5b1-2f4d9ded1f5a.png)
Part 3
![Screenshot 2021-09-10 at 13 17 29](https://user-images.githubusercontent.com/14430883/132845718-97aebc48-d089-4aef-9a2a-41f6eab4a696.png)

nezdolik commented:
oh i see a typo in second graph (endoint->endpoint), need to regenerate and re-upload the graph

nezdolik commented:
@tonya11en this is ready for review, meanwhile i will test slow start patch in real life (our infra)

tonya11en commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -833,6 +882,32 @@ HostConstSharedPtr EdfLoadBalancerBase::chooseHostOnce(LoadBalancerContext* cont
+  if (host_create_duration < slow_start_window_ &&
+      host.health() == Upstream::Host::Health::Healthy) {
+    aggression_ = aggression_runtime_ != absl::nullopt ? aggression_runtime_.value().value() : 1.0;
+    aggression_ = std::max(0.0, aggression_);
```
If the aggression somehow was set to be < 0, can you throw in a ENVOY_LOG_EVERY_POW_2 and log an error when this happens?

mattklein123 replied:
Per above I think we can remove this error handling here?

nezdolik replied:
we should keep it, as we cannot do validation in PGV layer

tonya11en commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -397,14 +400,22 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
+  double applySlowStartFactor(double host_weight, const Host& host);
+
 private:
+  friend class EdfLoadBalancerBasePeer;
```
What's going on here?

nezdolik replied:
i use this pattern to assert low level impl details in tests, eg config values for slow start and latest host added time, gives more confidence.

nezdolik replied:
somehow all my replies/comments did not persist, so will duplicate them. I use this pattern to assert low level impl details in tests, like values for slow start config options and latest host added time, gives more confidence.

tonya11en commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -542,16 +585,21 @@ class LeastRequestLoadBalancer : public EdfLoadBalancerBase,
     if (active_request_bias_ == 1.0) {
-      return static_cast<double>(host.weight()) / (host.stats().rq_active_.value() + 1);
+      host_weight = static_cast<double>(host.weight()) / (host.stats().rq_active_.value() + 1);
+    } else if (active_request_bias_ != 0.0 && active_request_bias_ != 1.0) {
```
`active_request_bias_` can't be 1.0 here because it would have made the `if` statement true?

tonya11en reviewed (commented):
Sorry for the delay! The new formula looks great and makes sense to me. Some minor comments are spread about, but looks good to me overall.

tonya11en commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,60 @@
+===============
+
+Slow start mode is a configuration setting in Envoy to progressively increase amount of traffic for newly added upstream endpoints.
+With no slow start enabled Envoy would send a proportional amount of traffic to new upstream ednpoints.
```
Can you fix the spelling in this doc as well, please. I'm not sure why the build didn't catch this.. I thought it used to.

nezdolik replied:
was surprising to me as well, perhaps spell check does not run on new .rst files

nezdolik replied:
I was also surprised that spell check did not catch those, perhaps it does not cover new .rst files...Fixed (with online spell checker)

tonya11en commented:
/wait

nezdolik commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -714,20 +733,37 @@ void EdfLoadBalancerBase::initialize() {
 
+void EdfLoadBalancerBase::recalculateHostsInSlowStart(const HostVector& hosts) {
+  auto current_time = time_source_.monotonicTime();
+  for (const auto& host : hosts) {
+    auto host_create_duration =
+        std::chrono::duration_cast<std::chrono::milliseconds>(current_time - host->creationTime());
+    // Check if host existence time is within slow start window.
+    if (host->creationTime() > latest_host_added_time_ &&
+        host_create_duration <= slow_start_window_ &&
+        host->health() == Upstream::Host::Health::Healthy) {
+      latest_host_added_time_ = host->creationTime();
+    }
+  }
```
We decided to fall back to simpler version of slow start algorithm to reduce code complexity. In previous version there was an ordered hash map struct with constant lookup that also required compaction with time (to purge stale hosts). I think is fine for initial version and would like to revisit it in case there is a perf penalty. Added a TODO.

nezdolik replied:
We have agreed to switch to simpler version of slow start algorithm to reduce code complexity. The previous version was maintaining ordered hash map to track hosts in slow start and required compaction as time progressed (to purge stale hosts). I think is fine to have this unoptimised linear scan in initial version, but would like to revisit in case there is a noticeable perf penalty. Added TODO

nezdolik commented:
@tonya11en thanks, addressed the review comments. I think only test code is left to review (unless it's already reviewed)

tonya11en reviewed (approved):
This LGTM! I like the ramp-up formula, I like the docs, I like the way it can be configured. This feature is going to be awesome. Thanks for all of this.

I'm holding off on that WRSQ patch https://github.com/envoyproxy/envoy/pull/18020 so we can get this merged without any more issues.

nezdolik commented:
Thanks for all the help and review @tonya11en @mattklein123 and community! I will wait for final review

mattklein123 commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -337,6 +337,35 @@ message Cluster {
+    // This parameter controls the speed of traffic increase over the slow start window. Defaults to 1.0,
+    // so that endpoint would get linearly increasing amount of traffic.
+    // When increasing the value for this parameter, the speed of traffic ramp-up increases non-linearly.
+    // The value of aggression parameter should be greater than 0.0.
```
I think you can add a PGV validation for this?

nezdolik replied:
does not look like PGV supports Envoy's `core.v3.RuntimeDouble` type

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,60 @@
+.. _arch_overview_load_balancing_slow_start:
```
We need release notes for the general feature and all of the new fields.

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,60 @@
+This could be undesirable for services that require warm up time to serve full production load and could result in request timeouts, loss of data and deteriorated user experience.
+
+Slow start mode is a mechanism that affects load balancing weight of upstream endpoints and can be configured per upstream cluster.
+Currently, slow start is supported in Round Robin and Least Request load balancer types.
```
nit: can you ref link to the relevant fields for each type?

mattklein123 commented on docs/root/intro/arch_overview/upstream/load_balancing/slow_start.rst:
```diff
@@ -0,0 +1,60 @@
+
+To reiterate, endpoint enters slow start mode:
+  * If no active healthcheck is configured per cluster, immediately if its cluster membership duration is within slow start window.
+  * In case an active healthcheck is configured per cluster, when its cluster membership duration is within slow start window and endpoint has passed an active healthcheck.
```
Sorry not sure where we landed on this, but wouldn't it make sense in this case to start slow start once the first active health check has passed? cc @tonya11en 

I would be fine if this were a future enhancement but just curious if we discussed this previously.

tonya11en replied:
That does make a lot more sense.. it just never occurred to me when I looked at it.

I'm in favor of starting the window upon the first successful HC, but given how long this PR has dragged on I think it should be in a subsequent diff. If we're in agreement, then I think a simple issue and link to this comment thread should suffice to track it.

mattklein123 replied:
Yes agree let's land and iterate. An issue is fine.

nezdolik replied:
+1 on opening a new issue (will do it tomorrow). also feeling a little bit exhausted from this pr taking so long

mattklein123 replied:
This issue still needs to be opened please or if it was link it here?

nezdolik replied:
Issue link: https://github.com/envoyproxy/envoy/issues/18351

mattklein123 commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -754,17 +755,36 @@ const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts
+                             ? std::chrono::milliseconds(DurationUtil::durationToMilliseconds(
+                                   slow_start_config.value().slow_start_window()))
+                             : std::chrono::milliseconds(0)),
+      aggression_runtime_(
```
This is no longer runtime and is a config value? Can you update this variable name? Also, you have error logging below which I think can also be caught now at config load time and/or PGV validation so that can be removed?

nezdolik replied:
it is a runtime double

mattklein123 replied:
I see OK. I guess that's why we cant' have PGV, OK got it.

mattklein123 commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -754,17 +755,36 @@ const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts
+              ? absl::optional<Runtime::Double>({slow_start_config.value().aggression(), runtime})
+              : absl::nullopt),
+      time_source_(time_source),
+      slow_start_enabled_(slow_start_window_ > std::chrono::milliseconds(0)),
```
nit: I don['t think this member variable is needed. I would just have a small helper function that does this calculation.

mattklein123 commented on source/extensions/clusters/redis/redis_cluster.h:
```diff
@@ -282,6 +282,8 @@ class RedisCluster : public Upstream::BaseDynamicClusterImpl {
   const Common::Redis::ClusterRefreshManagerSharedPtr refresh_manager_;
   const Common::Redis::ClusterRefreshManager::HandlePtr registration_handle_;
+
+  TimeSource& time_source_;
```
Why are all these changes needed to the other load balancers? Can this all be removed?

nezdolik replied:
yes, good catch

mattklein123 reviewed (changes requested):
Thanks very excited to see this land. Some random small comments.

/wait

nezdolik commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -1,3 +1,4 @@
+#include <functional>
```
nit: this is part of previous implementation version, should be removed

nezdolik commented:
Applied the latest review comments, fingers crossed that all checks will pass

nezdolik commented:
@mattklein123 am not sure why `envoy-presubmit` failed (looking at the logs it says operation was cancelled)

mattklein123 reviewed (changes requested):
Awesome, thanks. Just one small comment and then let's ship!

/wait

nezdolik commented on test/integration/stats_integration_test.cc:
```diff
@@ -268,8 +268,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSize) {
   // 2020/09/11  12973                38993   upstream: predictive preconnect
   // 2020/10/02  13251                39326   switch to google tcmalloc
-  // 2021/08/15  17290                40349   add all host map to priority set for fast host
-  //                                          searching
```
fixed

mattklein123 reviewed (approved):
Thanks, awesome work!

tonya11en commented:
🎉 🎉 🎉

wbpcode commented:
🎉 🎉 🎉 It's great.

costimuraru commented:
Woohoo!

du2016 commented:
🎉 🎉 🎉

ramaraochavali commented:
@nezdolik  one question on this - I understand that for new deployments, when all the pods are in slow start mode, all of them receive [similar amount based on their host weights](https://github.com/envoyproxy/envoy/blob/008c6d7b3cb4239607d68a4578d45f386c9c6775/source/common/upstream/load_balancer_impl.cc#L941) - so slow start mode is essentially not useful in that case and mostly would make sense if new pods come in HPA case. Is that correct?

nezdolik commented:
> @nezdolik one question on this - I understand that for new deployments, when all the pods are in slow start mode, all of them receive [similar amount based on their host weights](https://github.com/envoyproxy/envoy/blob/008c6d7b3cb4239607d68a4578d45f386c9c6775/source/common/upstream/load_balancer_impl.cc#L941) - so slow start mode is essentially not useful in that case and mostly would make sense if new pods come in HPA case. Is that correct?

Correct @ramaraochavali

ramaraochavali commented:
Thank you
